src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php

```diff
@@ -174,12 +178,24 @@ private function checkType(Definition $checkedDefinition, $value, \ReflectionPar
             throw new InvalidParameterTypeException($this->currentId, $class, $parameter);
         }
 
-        if ($value instanceof Parameter) {
-            $value = $this->container->getParameter($value);
-        } elseif ($value instanceof Expression) {
+        if ($value instanceof Expression) {
             $value = $this->getExpressionLanguage()->evaluate($value, ['container' => $this->container]);
-        } elseif (\is_string($value) && '%' === ($value[0] ?? '') && preg_match('/^%([^%]+)%$/', $value, $match)) {
-            $value = $this->container->getParameter($match[1]);
+        } elseif (\is_string($value)) {
+            if ('%' === ($value[0] ?? '') && preg_match('/^%([^%]+)%$/', $value, $match)) {
+                // Only array parameters are not inlined when dumped.
+                $value = [];
+            } elseif ($envPlaceholderUniquePrefix && false !== strpos($value, 'env_')) {
+                // If the value is an env placeholder that is either mixed with a string or with another env placeholder, then its resolved value will always be a string, so we don't need to resolve it.
+                // We don't need to change the value because it is already a string.
+                if ('' === preg_replace('/'.$envPlaceholderUniquePrefix.'_\w+_[a-f0-9]{32}/U', '', $value, -1, $c) && 1 === $c) {
```

Comment:
3 cases here (with `param: "%env(FOO)%"`:
- only 1 env placeholder (eg: `%param%`) -> we try to resolve it to get its type.
- more than 1 chained env placeholder (eg: `%param%%param%`) -> it has to be a string
- mixed env placeholder with a string (eg : `foo%param%`) -> it has to be a string